File: PDB/listSpliced.js
#!/usr/bin/env node

const argv = require('yargs').argv;
const _ = require('lodash');
const utils = require('./lib/utils');

const filesDir = argv._[0];

// Filename = pdbCode of a given protein.
// key = one <pdbCode>.txt = interface residues of <pdbCode>
// value = content of <pdbCode>.txt
// One line in the files = "Chain Resi Resn" (Residue index, Residue type)
const data = {};

function createInterfaceResiduesFromFile(filename, content) {
    const lines = content
        // Remove CR for line endings in Windows.
        .replace(/\r/g, '')
        .split('\n')
        .filter(line => !!line)
        .map(line => line.split(' '));

    const residuesAnalysis = lines.reduce((chains, residue) => {
        const chain = residue[0];

        if (!chains[chain]) {
            chains[chain] = {
                residues: [],
                sequences: [],
            };
        }

        chains[chain].residues.push({
            resi: parseInt(residue[1], 10),
            resn: utils.getOneLetterCode([residue[2]]),
        });

        return chains;
    }, {});

    Object.keys(residuesAnalysis).forEach((chain) => {
        const rawSequences = residuesAnalysis[chain].residues.reduce((seqs, residue, index) => {
            const lastIndex = index === 0 ? 0 : seqs.length - 1;
            const lastSeq = seqs[lastIndex];

            if (lastSeq && lastSeq[lastSeq.length - 1].resi === residue.resi - 1) {
                lastSeq.push(residue);
            } else {
                const newSeq = [residue];

                seqs.push(newSeq);
            }

            return seqs;
        }, []);

        residuesAnalysis[chain].sequences = _.sortBy(rawSequences.filter(seq => seq.length > 1), seq => seq.length).reverse();
    });

    return residuesAnalysis;
}


utils.readFiles(filesDir, (filename, content) => {
    const pdbCode = filename.replace('.txt', '');
    const residues = createInterfaceResiduesFromFile(filename, content);

    data[pdbCode] = residues;

    // console.log(JSON.stringify(data, null, 4));

    utils.clustalOmegaOutput(pdbCode, residues);
}, (err) => {
    throw err;
});

// First script: ???
// Second script: statistics on resn
// Third script: output amino acid sequences with "x" in gaps

File: PDB/list_spliced.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from lib import utils

STORE_DIR = 'store'
DATA_FILE = 'data.json'


def generate_output(data):
    '''
    Generate output in Clustal Omega format for our data structure.
    The clustal omega format has two lines: a (unique) header to identify
    the sequence, and the sequence itself.
    The sequence is a succession of one-letter amino acid codes.
    '''
    output = []

    for code in data:
        for interface in data[code]['interfaces']:
            residues = data[code]['interfaces'][interface]['residues']

            for chain in interface:
                sequences = []

                for i, r in enumerate(residues):
                    # TODO Save as number upfront in JSON
                    resi = int(r['resi'])
                    if r['chain'] == chain:
                        prev_i = len(sequences) - 1
                        prev_seq = sequences[prev_i] if prev_i > 0 else []
                        # print i, prev_i, prev_seq, resi

                        # If the last resi of the sequence is current resi - 1...
                        if prev_seq and int(prev_seq[len(prev_seq) - 1]['resi']) == resi - 1:
                            prev_seq.append(r)
                        else:
                            sequences.append([r])

                order = 0
                for seq in sequences:
                    if len(seq) > 1:
                        output.append((
                            utils.get_clustal_omega_header({
                                'code': code,
                                'name': data[code]['name'].replace(' ', ''),
                                'interface': interface,
                                'chain': chain,
                                'suffix': '_' + str(order),
                            }),
                            ''.join([utils.get_one_letter_code(r['resn'], code) for r in seq])
                        ))
                        order += 1

    return output


def main():
    data = utils.load_json(STORE_DIR + '/' + DATA_FILE)
    output = generate_output(data)

    output.sort(key=lambda pair: len(pair[1]), reverse=True)

    for (header, sequence) in output:
        print header.encode('utf-8')
        print sequence.encode('utf-8')


if __name__ == "__main__":
    main()
